push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java

```diff
@@ -169,4 +188,42 @@ public void loadAndQueueTokenBatch(@Observes @Dequeue MessageHolderWithVariants
             }
         }
     }
+
+    /**
+     * Tries to submit token batch to queue. Returns true in case of success.
+     *
+     * It retries submitting when it detects that queue is full until it times out and return false;.
+     *
+     * @return returns true if token batch was queued; return false is queue was still full even after retrying
+     */
+    private boolean tryToDispatchTokens(MessageHolderWithTokens msg) throws InterruptedException {
+        for (int i = 0 ; i <= MAX_TIME_TO_BLOCK_WHEN_QUEUE_IS_FULL; i += TIME_BETWEEN_RETRIES_WHEN_QUEUE_IS_FULL) {
+            if (dispatchTokensOrFailWhenQueueIsFull(msg)) {
+                return true;
+            } else {
+                logger.fine(String.format("Queue is full, blocking loading of batch token #%s for %s variant (%s) for %s ms", msg.getSerialId(), msg.getVariant().getType().getTypeName(), msg.getVariant().getVariantID(), TIME_BETWEEN_RETRIES_WHEN_QUEUE_IS_FULL));
+                Thread.sleep(TIME_BETWEEN_RETRIES_WHEN_QUEUE_IS_FULL);
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Tries to dispatch tokens; returns true if tokens were successfully queues.
+     * Detects when queue is full and in that case returns false.
+     *
+     * @return returns true if tokens were successfully queued; returns false if queue was full
+     */
+    private boolean dispatchTokensOrFailWhenQueueIsFull(MessageHolderWithTokens msg) {
+        try {
+            dispatchTokensEvent.fire(msg);
+            return true;
+        } catch (MessageDeliveryException e) {
+            Throwable cause = e.getCause();
+            if (cause.getMessage() != null && cause.getMessage().contains("is full")) {
```

Comment:
yea, we would have to go implementation-specific, Hornetq/ActiveMQ use different classes